tket/src/Circuit/Boxes.cpp

```diff
@@ -285,14 +285,37 @@ bool CompositeGateDef::operator==(const CompositeGateDef &other) const {
   return this->get_def()->circuit_equality(*other.get_def(), {}, false);
 }
 
+static const composite_def_ptr_t &get_non_null_ptr(
+    const composite_def_ptr_t &gate) {
+  if (gate) {
+    return gate;
+  }
+  throw std::runtime_error(
+      "Null CompositeGateDef pointer passed to CustomGate");
+}
+
 CustomGate::CustomGate(
     const composite_def_ptr_t &gate, const std::vector<Expr> &params)
-    : Box(OpType::CustomGate, gate->signature()), gate_(gate), params_(params) {
+    : Box(OpType::CustomGate, get_non_null_ptr(gate)->signature()),
+      gate_(gate),
+      params_(params) {
   if (params_.size() != gate_->n_args()) throw InvalidParameterCount();
 }
 
 CustomGate::CustomGate(const CustomGate &other)
-    : Box(other), gate_(other.gate_), params_(other.params_) {}
+    // Should be impossible to get a null gate ptr, but check anyway.
```

Comment:
If it should be impossible then we should `TKET_ASSERT`, not throw. (But I'm not sure even that is worthwhile.)

That also means we only call `get_non_null_ptr` in one place, so it doesn't need to be a separate function and the check should just be integrated into the other constructor.

Reply:
It's impossible currently, but as soon as someone writes a new constructor and forgets to ensure that gate_ is not null, it will become possible again (admittedly unlikely, since I hope they would look at the current constructor and see the problem).

Perhaps we should consider using a not_null template?

Reply:
Let's cross that bridge when we come to it. In any case the copy constructor should not have to repeat validation of the data.